RFC: Add documentation page about error handling and logging.

hlinnaka opened the pull request:
This PR adds a page to the internal documentation, on how we do error handling and logging. I'm adding it to the main docs, rather than the `rfcs` section, because this is something that we would continue to update when things change. But this hasn't been discussed much before, so I am Requesting Comments on it.

@danieltprice this is just for internal developer docs, not the user-facing docs, but maybe you have some good advice here too.

@lassizci can you add or suggest something here about the json logging? Tips on how to process the json format etc.

hlinnaka commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+For example, if deleting a file fails because the file already didn't exist,
+it should be logged as Warning.
+
+> **Note:** The python regression tests, under `test_regress`, check the
+> pageserver log after each test for any ERROR and WARN lines. If there are
+> any ERRORs or WARNs that have not been explicitly listed in the test as
+> allowed, the test is marked a failed. This is to catch unexpected errors
+> e.g. in background operations, that don't cause immediate misbehaviour in
+> the tested functionality.
```
This is not in 'main' yet. It is being done in https://github.com/neondatabase/neon/pull/2653.

lassizci commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+            // of the day isn't very important though, so we just WARN and
+            // continue with the default in any case.
+            if err.kind() != std::io::ErrorKind::NotFound {
+                 tracing::warn!("could not read \"/etc/motd\": {err:?}");
```
Instead of just formatting the message, it is helpful for generating structured logs when relevant bits of information are recorded to separate fields. This particular example might not be the prime one for that, but here we could do:
```rust
warn!(file = "/etc/motd", "could not read {file}: {err:?}");
```

So the log output becomes(plain format):
```console
2022-10-23T15:09:32.007339Z  WARN could not read /etc/motd: PermissionDenied file=/etc/motd
```

and with json formatting we would have:
```json
{"timestamp":"2022-10-23T15:09:32.007339Z","level":"WARN","fields":{"message":"could not read /etc/motd: PermissionDenied","file":"/etc/motd"}}
```

We could distill a couple sentences about structuring and then have our examples follow that.

koivunej replied:
Re: my other https://github.com/neondatabase/neon/pull/2681#issuecomment-1288554825, I'd go as far as not including any variable context in the message:

```rust
warn!(?file, ?err, "read failed");
```

But will need to do some experimentation on what this would look like.

lassizci replied:
I would like to agree and partially do, but wouldn't suggest completely against having some context in the message either, because it's a bit less tool agnostic in sense `message` field becomes a bit less useful and then there needs to be a view for each type of logs. So would write the message for humans and separate the fields for computers, so that they can make things more accessible for humans.

sharnoff commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+- could not remove temporary file "foobar.temp"
+- unrecognized file "foobar" in timeline directory
+
+Warnings are similar to Errors, in that they should not happen
+when the system is operating normally. The difference between Error and
+Warning is that an Error means that the operation failed, whereas Warning
+means that something unexpected happened, but the operation continued anyway.
```
I really like this definition. It aligns with what I've intuitively been doing for a while.

sharnoff commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+frequently when compute nodes are shut down, or are killed abruptly
+because of errors in the compute.
+
+**Errors are monitored, and always need human investigation to determine
+the cause.**
```
Is this currently the case? I'd assume because of #2653 that it's not

hlinnaka replied:
They are not monitored currently. This sentence is aspirational :-).

sharnoff commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+
+### Debug & Trace
+
+Debug and Trace level messages are not Debug-level messages are not
+printed to the log in our normal production configuration, but could
+be enabled for a specific server or tenant, to aid debugging.
+(Although we don't actually have that capability as of this writing).
```
It'd be good to have more information on what the distinction between Debug and Trace should be, and when to use them instead of Info

lassizci replied:
There was a brief discussion whether we should adopt https://github.com/trentm/node-bunyan#levels since there would have been a rust library following the same scheme. Although didn't necessarily warm up to idea to introduce one more level, because that would mean going through pretty much any existing error to figure out if it's error or fatal, the bunyan definitions could have some useful points.

hlinnaka replied:
I like those node-bunyan definitions, and an additional FATAL level also makes sense to me. However, `tracing` doesn't have FATAL, so I don't think we can adopt that.

sharnoff reviewed (commented):
Glad to see an RFC about this written up! I have a couple general thoughts, and a handful of comments that largely reference these. In general, I agree with the sentiments; comments are largely "is it worth considering XYZ?".

---

In short: IMO there are two (somewhat) opposed models for error handling + logging:

1. Span-based -- spans should be used instead of `context`, and errors either (a) collect span info on creation \[not sure how to do this\] or (b) are logged at point of creation so they include that span info. Spans are always constructed, which means we *do* pay a performance penalty (although maybe this doesn't _actually_ matter).
2. `.context()`-based -- spans are not used as much, and errors are logged at point of consumption with additional information provided by `.context` (for `anyhow` errors, or the equivalent). Context is only constructed when errors happen, so it's "zero cost" for happy paths.

However, there's a problem with this: In general, we want span-based logging for `info!` (and even `error!`) in order to take advantage of structured-logging capabilities, but we want `context`-based error handling so that we can decide the significance of an error *after* it happens.

I don't really know how best to reconcile these, but I think it's worth considering. There's a hybrid approach I discussed in one of the comments below; it solves some problems, and creates others.

---

As a final note: It's relatively simple to build on `anyhow`! (or `eyre`, an `anyhow` fork with customization) We don't _need_ to, but there's a lot of "works _juuuust_ right" we can get out of a custom `anyhow` replacement. I think it's worth considering; some of my suggestions are of the form "if we had a custom error type, we could solve this by XYZ". There's also definitely a cost to using our own set of error types (initial conversion, onboarding, docs, etc.), so it's meant primarily as a hypothetical "here's some things we could do, if had them".

koivunej commented:
> 1. Span-based -- spans should be used instead of context, and errors either (a) collect span info on creation [not sure how to do this] or (b) are logged at point of creation so they include that span info. Spans are always constructed, which means we do pay a performance penalty (although maybe this doesn't actually matter).

In the collected logs, I find the stacktraces disorienting. Wanting to bring it up somehow in the context of this RFC, this comment immediatedly lead me to thinking of [`SpanTrace`](https://docs.rs/tracing-error/latest/tracing_error/struct.SpanTrace.html) instead of stacktrace logging. Wouldn't that help to leave our `anyhow!`s clean of context, and just describe what failed? It looks like eyre (or color-eyre) would have support for `SpanTrace`s.

hlinnaka commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+
+## Logging errors
+
+The principle is that errors are logged when they are handled. If you
+just propagate an error to the caller in a function, you don't need to
+log it; the caller will. But if you "swallow" an error in a function,
+you *must* log it (if it needs to be logged at all).
```
> As I mentioned above, this has the tendency to divorce errors from the context of the spans they were created in.

Huh, you're right! For some reason, my mental model has been that an anyhow::Error captures the span that it was created in, but it doesn't.

There is https://docs.rs/tracing-error/latest/tracing_error/ for capturing the current span, but anyhow won't use it automatically.

koivunej replied:
Raced to comment on `SpanTrace` in topmost thread: https://github.com/neondatabase/neon/pull/2681#issuecomment-1288554825 but this thread seems better suited.

After some [experimentation](https://gist.github.com/koivunej/6f6ac896ddea7f6be01245bb2ca56097) (includes example output in comment), `color-eyre` is required to format and `tracing-error` are required to capture the `SpanTrace`. The SPANTRACE header comes from [`color-spantrace`](https://github.com/yaahc/color-spantrace/blob/1a52aac2b55a05efc64807f06b91414550805af8/src/lib.rs#L353).

sharnoff replied:
That's very cool! Seeing that, I wonder if it's possible to capture the `Span` object itself, so that logging errors can temporarily re-enter all of the necessary spans instead. If not, I think `SpanTrace` is still a great solution to this :)

koivunej commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+
+### Debug & Trace
+
+Debug and Trace level messages are not Debug-level messages are not
```
```suggestion
Debug and Trace level messages are not
```

SomeoneToIgnore commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+responds with the HTTP 403 Already Exists error in that case, rather
+than a generic HTTP 500 Internal Server Error.
+
+If you need to distinguish between different kinds of errors, create a
+new `Error` type. The `thiserror` create is useful for that. But in
+most cases `anyhow::Error` is good enough.
```
I practice poor man's typing with anyhow 

https://github.com/neondatabase/neon/blob/9f49605041cc4954eff96690c7874cc16ac3f8f8/libs/remote_storage/src/lib.rs#L154-L162

which can indicate various failure scenarios with the enum variants.

problame replied:
See @funbringer 's comment https://github.com/neondatabase/neon/pull/2681/files#r1004581973

SomeoneToIgnore commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+
+It is acceptable to use functions that might panic, like `unwrap()`, if
+it is obvious that it cannot panic. For example, if you have just
+checked that a variable is not None, it is OK to call `unwrap()` on it.
```
```suggestion
checked that a variable is not None, it is OK to call `unwrap()` on it, but still prefer to use `expect("reason")` instead to explain why it cannot fail.
```

danieltprice replied:
"but it is still preferable  to use `expect("reason")` instead to explain why the function cannot fail."

hlinnaka replied:
thanks, added!

SomeoneToIgnore reviewed (commented):
I think it's worth explaining anyhow error formatters (`{:?}` vs `{:#}` vs other options and when we want to use either)

funbringer commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+
+## Error types
+
+We use the `anyhow` crate widely. It contains many convenient macros
```
I have a different interpretation which involves two layers of thinking:

* On the surface, `.context(..)` [is](https://docs.rs/anyhow/latest/src/anyhow/context.rs.html#90-105) `ok_or_else(|| anyhow!(...))`. Which means that every time I see the latter, the question pops up: "why this again?". The name _might_ (but read below) be controversial, but the method itself is convenient.

* > it's unclear when reading Option::context that it's the source of the error instead of something providing context to an existing error

Oh, but there's an error and it's `None`! This is the source we've been looking for (although it's very shy and prefers to keep its secrets). The method does what it claims to do: it attaches a context to _an error_. Even the `eyre`'s doc explicitly says that: _anyhow avoids this issue by using a **different mental model** where you’re adding “context” to an error, though this not a mental model for error handling that eyre agrees with._

Clearly, you're accustomed to the `eyre`'s model, but this doesn't necessarily mean that the other model is wrong. Especially considering the first bullet.

sharnoff replied:
> Clearly, you're accustomed to the `eyre`'s model, but this doesn't necessarily mean that the other model is wrong.

Fair enough :)

> Oh, but there's an error and it's `None`!

I think this is the part that I disagree with, but if I'm in the minority here that's ok

LizardWizzard commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+
+If you need to distinguish between different kinds of errors, create a
+new `Error` type. The `thiserror` create is useful for that. But in
+most cases `anyhow::Error` is good enough.
```
It seems that I'm in minority here. But I'll raise my hand anyway :)

To me error handling with anyhow is broken for anything other than simple CLIs which just bubble up errors to the user without trying to inspect them. 

The ideal approach to me is that every module defines its error enum with `thiserror::Error` (it just saves boilerplate on writing impls manually). It has the benefit that anyone who sees this module for the first time can see what are the possible errors. Listing error variants requires to put more thought into it compared to just a bunch of `bail!("A is broken")` calls. Better names for errors lead to better answers to "How I can handle this error?" type of questions. 

Nesting anyhow::Error types in other enum has following downsides:
* When there are many anyhow::Error variants `From` doesnt work anymore which leads to boilerplate since `?` is not available.
* If you need to detalize some anyhow::Error into new variants then you have to change every caller of a function returning such an error. With thiserror you need only add new variant and change the places which actually handle the error

This approach is more radical and requires ground up design for error types which is not an easy task

danieltprice commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+operation. Almost every operation happens on a particular tenant and
+timeline, so we enter a span with the "tenant_id" and "timeline_id"
+very early when processing an incoming API request, for example. All
+background operations should also run in a span contain at least those
```
```suggestion
background operations should also run in a span containing at least those
```

danieltprice commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+
+The principle is that errors are logged when they are handled. If you
+just propagate an error to the caller in a function, you don't need to
+log it; the caller will. But if you "swallow" an error in a function,
```
"consume" might be a better term than "swallow"

danieltprice commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+caller, and it is up to the caller to log the error or propagate it
+further, if the failure is not expected. But if, for example, it is
+normal that the "/etc/motd" file doesn't exist, the caller can choose
+to silently ignore the error, or log it in INFO or DEBUG level:
```
```suggestion
to silently ignore the error, or log it as an INFO or DEBUG level message:
```

danieltprice commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+between different error cases. Most errors are propagated all the way
+to the mgmt API handler function, or the main loop that handles a
+connection with the compute node, and they are all handled the same
+way: the error is logged and returned the error to the client as
```
```suggestion
way: the error is logged and returned to the client as
```

danieltprice commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+than a generic HTTP 500 Internal Server Error.
+
+If you need to distinguish between different kinds of errors, create a
+new `Error` type. The `thiserror` create is useful for that. But in
```
```suggestion
new `Error` type. The `thiserror` crate is useful for that. But in
```

danieltprice commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+
+These should *not* happen during normal operations. "Normal
+operations" is not a very precise concept. But for example, disk
+errors is not expected to happen when the system is working, so that
```
```suggestion
errors are not expected to happen when the system is working, so those
```

danieltprice commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+These should *not* happen during normal operations. "Normal
+operations" is not a very precise concept. But for example, disk
+errors is not expected to happen when the system is working, so that
+counts as an Error. However, if a TCP connection to a compute node is
```
```suggestion
count as Errors. However, if a TCP connection to a compute node is
```

danieltprice commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+
+Whether something should be logged at ERROR, WARNING or INFO level can
+depend on the callers and clients. For example, it might be unexpected
+and a sign of a serious issue, if the console calls the
```
```suggestion
and a sign of a serious issue if the console calls the
```

danieltprice commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+"timeline_detail" mgmt API for a timeline that doesn't exist. ERROR
+would be appropriate in that case. But if the console routinely calls
+the API after deleting a timeline, to check if the deletion has
+completed, then it would be totally normal and INFO or DEBUG level
```
```suggestion
completed, then it would be totally normal and an INFO or DEBUG level message
```

danieltprice commented on docs/error-handling.md:
```diff
@@ -0,0 +1,190 @@
+Follow that guide when writing error messages in the PostgreSQL
+extension. We don't follow it strictly in the pageserver and
+safekeeper, but the advice in the PostgreSQL style guide is generally
+good, and you can't go wrong by following it.
```
+1
Microsoft also provides decent error message writing guidelines and best practices. https://learn.microsoft.com/en-us/windows/win32/debug/error-message-guidelines
We might also give some thought to what an error message review process would look like. I have found such reviews to occur upon request, periodically, or before a feature is released. I am not sure what approach makes the most sense for Neon, but I would be interested in any thoughts on this.

hlinnaka replied:
> Microsoft also provides decent error message writing guidelines and best practices. https://learn.microsoft.com/en-us/windows/win32/debug/error-message-guidelines

Good advice

> We might also give some thought to what an error message review process would look like. I have found such reviews to occur upon request, periodically, or before a feature is released. I am not sure what approach makes the most sense for Neon, but I would be interested in any thoughts on this.

For the storage components, I think any PR that includes error messages is the right place to review the messages too. A separate review would draw more attention to error messages and make them better, for sure, but I don't think it's worth the effort. (For the user-facing web UI and APIs, error messages are much more prominent and more attention might be warranted.)

problame commented:
@hlinnaka I think there are a bunch of suggested edits that you can apply to declutter this.

hlinnaka commented:
I applied the minor copy edits that y'all suggested, thanks!

There are three discussions here that haven't been fully resolved:

- Using `anyhow` vs. defining fine-grained error types. I think the principle explained in this RFC is good, and that's what we've been doing so far. However, there's some disagreement on that. If we want to change that, let's discuss that as a separate RFC. (https://github.com/neondatabase/neon/pull/2681#discussion_r1004328710)

- Capturing spans vs stacktraces in errors. I think we're all more or less in agreement on how we wish spans would work, but we're not there yet. So I just added a little Todo comment to note on that, and that we might start using SpanTrace in the future. (https://github.com/neondatabase/neon/pull/2681#issuecomment-1288554825)

- Using variables in error messages vs. structured logging (https://github.com/neondatabase/neon/pull/2681#discussion_r1002726914).

I think it's time to push this, Those things can be proposed and changed as separate RFCs and PRs, if we want to, but this document fairly accurately describes how we do things now.

hlinnaka commented:
> I think it's worth explaining anyhow error formatters (`{:?}` vs `{:#}` vs other options and when we want to use either)

+1. To be honest, it's not very clear to me either. Would you like to write up a short proposal? :-). It can be done as a separate PR.

sharnoff commented:
> > I think it's worth explaining anyhow error formatters ({:?} vs {:#} vs other options and when we want to use either)
>
> Would you like to write up a short proposal?

FWIW, there's a [section about this](https://docs.rs/anyhow/latest/anyhow/struct.Error.html#display-representations) in the `anyhow::Error` docs. TL;DR is:

* `{}` prints only the outermost context -- e.g., "Failed to X"
* `{:#}` prints all errors' messages one after another, with colons -- e.g., "Failed to X: Couldn't find Y: Missing key for Z"
* `{:?}` pretty-prints the same error messages, and also displays a backtrace if there is one - e.g.:
    ```
    Error: Failed to X
    
    Caused by:
        Couldn't find Y
        Missing key for Z
   
    Stack backtrace:
        <it's very long>
    ```
* `{:#?}` uses the type's `Debug` pretty-printing (includes indentation)

Worth noting: the "caused by" stuff uses the `cause` method on `std::error::Error` - it's possible for custom errors to play nice in the middle of the list of causes.

vadim2404 commented:
Surfing the internet, I found this library https://docs.rs/error-stack/latest/error_stack/

At least, we should take it into consideration

hlinnaka commented:
> Surfing the internet, I found this library https://docs.rs/error-stack/latest/error_stack/
> 
> At least, we should take it into consideration

Looks neat, might be a good alternative. It has the support for capturing spans, like SpanTrace, so that's good. And being able to attach "contexts" and "attachments" to errors seems useful. My main worry is the same as with `thiserror`: defining explicit error types for everything is much more effort than using anyhow, and in almost all cases, we handle all errors in the same way so there's not much benefit from the explicit error types. The docs on error_stack imply using it with fine-grained errors, but maybe it would be usable with anyhow too.

One risk is that it's a very new project, announced in June 2022. It might not become popular, and for something like error handling that's used all over the place, I would like to stick to one of the popular crates.
